npm/pkg/controlplane/goalstateprocessor/goalstateprocessor.go

```diff
@@ -0,0 +1,286 @@
+
+	klog.Infof("Creating GoalStateProcessor for node %s", nodeID)
+	return &GoalStateProcessor{
+		ctx:            ctx,
+		nodeID:         nodeID,
+		podName:        podName,
+		dp:             dp,
+		inputChannel:   inputChan,
+		backoffChannel: make(chan *protos.Events),
+	}
+}
+
+func (gsp *GoalStateProcessor) Start() {
+	klog.Infof("Starting GoalStateProcessor for node %s", gsp.nodeID)
+	go gsp.run()
+}
+
+func (gsp *GoalStateProcessor) SetController(controllerIP string, controllerPort int) {
+	klog.Infof("Setting controller for node %s", gsp.nodeID)
+	gsp.controllerIP = controllerIP
+	gsp.controllerPort = controllerPort
+}
+
+func (gsp *GoalStateProcessor) Stop() {
+	klog.Infof("Stopping GoalStateProcessor for node %s", gsp.nodeID)
+	gsp.cancel()
+}
+
+func (gsp *GoalStateProcessor) run() {
+	if gsp.controllerIP != "" && gsp.controllerPort != 0 {
+		klog.Warningf("Invaling controller for node %s, IP %s port %d", gsp.nodeID, gsp.controllerIP, gsp.controllerPort)
+		return
+	}
+	klog.Infof("Starting dataplane for node %s", gsp.nodeID)
+	for {
+		select {
+		case <-gsp.ctx.Done():
+			klog.Infof("GoalStateProcessor for node %s stopped", gsp.nodeID)
+			return
+		case inputEvents := <-gsp.inputChannel:
+			klog.Infof("Received event %s", inputEvents)
+			gsp.process(inputEvents)
+		case backoffEvents := <-gsp.backoffChannel:
+			// For now keep it simple. Do not worry about backoff events
+			// but if we need to handle them, we can do it here.
+			klog.Infof("Received backoff event %s", backoffEvents)
+			gsp.process(backoffEvents)
+		}
+	}
+}
+
+func (gsp *GoalStateProcessor) process(inputEvent *protos.Events) {
+	klog.Infof("Processing event")
+
+	// Process these individual buckkets in order
+	// 1. Apply IPSET
+	// 2. Apply POLICY
+	// 3. Remove POLICY
+	// 4. Remove IPSET
+	payload := inputEvent.GetPayload()
+
+	if !validatePayload(payload) {
+		klog.Warningf("Empty payload in event %s", inputEvent)
+		return
+	}
+
+	err := gsp.processIPSetsApplyEvent(payload[cp.IpsetApply])
+	if err != nil {
+		klog.Errorf("Error processing IPSET apply event %s", err)
+	}
+
+	err = gsp.processPolicyApplyEvent(payload[cp.PolicyApply])
+	if err != nil {
+		klog.Errorf("Error processing POLICY apply event %s", err)
+	}
+
+	err = gsp.processPolicyRemoveEvent(payload[cp.PolicyRemove])
+	if err != nil {
+		klog.Errorf("Error processing POLICY remove event %s", err)
+	}
+
+	err = gsp.processIPSetsRemoveEvent(payload[cp.IpsetRemove])
+	if err != nil {
+		klog.Errorf("Error processing IPSET remove event %s", err)
+	}
+}
+
+func (gsp *GoalStateProcessor) processIPSetsApplyEvent(goalState *protos.GoalState) error {
+	for _, gs := range goalState.GetData() {
+		payload := bytes.NewBuffer(gs)
+		ipset, err := cp.DecodeControllerIPSet(payload)
+		if err != nil {
+			return err
+		}
+
+		ipsetName := ipset.GetPrefixName()
+		klog.Infof("Processing %s IPSET apply event", ipsetName)
+
+		cachedIPSet := gsp.dp.GetIPSet(ipsetName)
```

Comment:
like @nitishm has said, it might make sense to make IPSet an interface since the fan-out controller doesn't need some fields in dp's ipset (e.g. reference counts, kernel counts, etc.)